dstep/translator/Translator.d

```diff
@@ -69,48 +70,62 @@ class Translator
         write(outputFile, translateToString());
     }
 
+    static size_t countTrailingEndls(string a) 
+    {
+        size_t itr = a.length;
+
+        while (itr != 0 && a[itr - 1] == '\n')
+            --itr;
+                
+        return a.length - itr;
+    }
+
+    static size_t countLeadingEndls(string a)
+    {
+        size_t count = 0;
+
+        while (count < a.length && a[count] == '\n')
+            ++count;
+
+        return count;
+    }
+
+    static void endlAwareAppend(ref Appender!string a, string b)
+    {
+        import std.algorithm.comparison;
+
+        size_t aTrailing = countTrailingEndls(a.data);
+        size_t bLeading = countLeadingEndls(b);
+        size_t bTrailing = countTrailingEndls(b);
+
+        size_t num = min(bLeading + aTrailing, 2);
+        a.put("\n".replicate(num < aTrailing ? 0 : num - aTrailing));
+        a.put(b[bLeading..$-bTrailing]);
+        a.put("\n".replicate(min(bTrailing, 2)));
+    }
```

Comment:
> So we would need to have some dstep AST (currently there are strings only) or use the clang AST but there may be no one to one mapping between dstep AST and clang AST.

In the case of DStep I imagine a function taking a cursor or a type which then does a if/switch and hard codes if a cursor type is multiline or not.

> What if there is a few nodes in one line?

This functionality is only for statements and other declarations that will end with a newline. For example, It's not used when translating the expressions in for loop. I basically have a "writer" for each lager component. Then a "translator" for something smaller like an expression. This functionality is only used between the "writers".

> I suppose it is the easiest to do if a full AST is known in advance.

Yes. But the case of my other project the library I'm using, which contains the AST, is using the visitor pattern. Then I'm visiting all nodes at a given level building up an array. Then I'm visiting all nodes in the array to do the actual processing. When the data structure is an array it's a lot easier to get the next and previous nodes.

This was all just an idea because I though the implementation was cleaner in my other project. But I have not looked into if that approach works in DStep.